### ISSUE TITLE
Initialising ITensors with data type T

### ISSUE THREAD

**Issue by jtschneider** [ITensor/ITensors.jl#197]
I was initially attempting to add the functionality to initialise an MPS with `CompleF64` datatype. 

So I went about and copied simply the `function MPS(sites)` as can be viewed in pull request https://github.com/ITensor/ITensors.jl/pull/196.

Then after testing I noticed that any `Int` data type is not recognised by the initialisation functions of an ITensor:
https://github.com/ITensor/ITensors.jl/blob/5f1f6422e95f07be6c0106e53ba7ea6d2e8e6651/src/itensor.jl#L99-L102
and
https://github.com/ITensor/ITensors.jl/blob/5f1f6422e95f07be6c0106e53ba7ea6d2e8e6651/src/itensor.jl#L111-L113

My question now: Is this conversion of the data type `T` to `float(T)` intended? It seems unexpected and might come surprising.

Cheers
Jan

**Comment by jtschneider** [comment]
This is indeed a recurring theme in the file [src/itensor.jl](https://github.com/ITensor/ITensors.jl/blob/master/src/itensor.jl). While most other initialising functions specifically mention this conversion of `T` to `float(T)`, those functions above and this one 
https://github.com/ITensor/ITensors.jl/blob/5f1f6422e95f07be6c0106e53ba7ea6d2e8e6651/src/itensor.jl#L127-L130
lack the explicit mentioning of the conversion of `T` to `float(T)`.

This seems to be a paradigm which might very well have already been discussed. I am very sorry if this is the case!

**Comment by mtfishman** [comment]
That is a good question, and it is worth revisiting these constructors. I think we should change the first case to not perform the `float(T)` conversion, so that `ITensor(Int,i,i')` really makes and ITensor with element type `Int`.

However, it may be nice to keep the second and third cases you list as is. The reason is that in most tensor network calculations, it is generally better to use floating point numbers since operations with them can be dispatched to BLAS, whereas matrices of Int use Julia's generic matrix multiplication algorithm:
```julia
julia> n = 1000; T = Float64; A = ones(T,n,n); B = copy(A); @btime $A*$B;
  2.353 ms (2 allocations: 7.63 MiB)

julia> n = 1000; T = Int64; A = ones(T,n,n); B = copy(A); @btime $A*$B;
  300.940 ms (8 allocations: 7.63 MiB)
```
For that reason, I think it is good to use the rule that ITensors have floating point element types, unless specified explicitly with something like `ITensor(Int,i,i')`. I think it is nice to just be able to use the syntax `ITensor([0 1; 1 0], i, i')` (for example for \sigma_x), and I think in general people would want that result to be an ITensor with a Float64 element type.

Definitely it would be good to update the docs about this functionality (whichever decision we ultimately make for these constructors).

**Comment by emstoudenmire** [comment]
Hi Jan, I would also like to ask: what is the use case you have in mind for initializing ITensors with an Int data type?

**Comment by jtschneider** [comment]
Hey Miles,
I cannot think of a practical application of an ITensor with integer coefficients...

**Comment by orialb** [comment]
> which means if the Julia base function float(z::Complex{Float}) = z where ever to not be the identity this behaviour would change

I think this is the intended behavior of `float`, and since changing this will be breaking it could not be changed in any future Julia 1.x version.

**Comment by jtschneider** [comment]
Yes, I think so, too.
I would be surprised by this breaking change in any future Julia version. With mentioning this, I merely wanted to archive this thought such that it could be useful in the unlikely event of such a change.

**Comment by emstoudenmire** [comment]
Thanks for looking carefully at this issue everyone

**Comment by mtfishman** [comment]
This is closed by #217.
